// @mongosh/cli-repl/lib/node-repl-fix-completer-side-effects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fixNodeReplCompleterSideEffectHandling = fixNodeReplCompleterSideEffectHandling;
const stream_1 = require("stream");
const util_1 = require("util");
let replHasNodeBug59774;
async function detectReplNodeBug59774() {
    const { start: replStart } = require('repl');
    const repl = replStart({
        input: new stream_1.PassThrough(),
        output: new stream_1.PassThrough(),
        terminal: true,
        useGlobal: false,
    });
    let ranFunction = false;
    repl.context.causeSideEffect = () => (ranFunction = true);
    try {
        await (0, util_1.promisify)(repl.completer)('causeSideEffect().x');
    }
    finally {
        repl.close();
    }
    return ranFunction;
}
async function fixNodeReplCompleterSideEffectHandling(completer) {
    replHasNodeBug59774 ??= await detectReplNodeBug59774();
    return async (text) => {
        const line = text.trimStart();
        if (!replHasNodeBug59774 ||
            /^\s*\.(\w*)$/.exec(line) !== null ||
            /\brequire\s*\(\s*['"`](([\w@./:-]+\/)?(?:[\w@./:-]*))(?![^'"`])$/.exec(line) !== null ||
            /\bimport\s*\(\s*['"`](([\w@./:-]+\/)?(?:[\w@./:-]*))(?![^'"`])$/.exec(line) !== null ||
            line.length === 0 ||
            /\w|\.|\$/.exec(line[line.length - 1]) === null) {
            return await completer(text);
        }
        const simpleMatch = /(?:[\w$'"`[{(](?:\w|\$|['"`\]})])*\??\.)*[a-zA-Z_$](?:\w|\$)*\??\.?$/.exec(line);
        if (simpleMatch !== null) {
            return await completer(simpleMatch[0]);
        }
        return [[], text];
    };
}
