// @mongosh/arg-parser/lib/arg-metadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsupportedArgumentError = exports.UnknownArgumentError = exports.InvalidArgumentError = exports.argMetadata = void 0;
exports.getArgumentMetadata = getArgumentMetadata;
exports.getDeprecatedArgsWithReplacement = getDeprecatedArgsWithReplacement;
exports.getUnsupportedArgs = getUnsupportedArgs;
const v4_1 = __importDefault(require("zod/v4"));
exports.argMetadata = v4_1.default.registry();
function getArgumentMetadata(schema, fieldName) {
    const fieldSchema = schema.shape[fieldName];
    if (!fieldSchema) {
        return undefined;
    }
    return exports.argMetadata.get(fieldSchema);
}
function getDeprecatedArgsWithReplacement(schema) {
    const deprecated = {};
    for (const fieldName of Object.keys(schema.shape)) {
        const meta = getArgumentMetadata(schema, fieldName);
        if (meta?.deprecationReplacement) {
            deprecated[fieldName] = meta.deprecationReplacement;
        }
    }
    return deprecated;
}
function getUnsupportedArgs(schema) {
    const unsupported = [];
    for (const fieldName of Object.keys(schema.shape)) {
        const meta = getArgumentMetadata(schema, fieldName);
        if (meta?.unsupported) {
            unsupported.push(fieldName);
        }
    }
    return unsupported;
}
class InvalidArgumentError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidArgumentError';
    }
}
exports.InvalidArgumentError = InvalidArgumentError;
class UnknownArgumentError extends Error {
    constructor(argument) {
        super(`Unknown argument: ${argument}`);
        this.name = 'UnknownArgumentError';
        this.argument = argument;
    }
}
exports.UnknownArgumentError = UnknownArgumentError;
class UnsupportedArgumentError extends Error {
    constructor(argument) {
        super(`Unsupported argument: ${argument}`);
        this.name = 'UnsupportedArgumentError';
        this.argument = argument;
    }
}
exports.UnsupportedArgumentError = UnsupportedArgumentError;
