// @mongosh/shell-api/lib/deep-inspect/service-provider-wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deepInspectServiceProviderWrapper = deepInspectServiceProviderWrapper;
const cursor_wrapper_1 = require("./cursor-wrapper");
const custom_inspect_1 = require("./custom-inspect");
function deepInspectServiceProviderWrapper(sp) {
    return {
        get [Symbol.for('@@mongosh.originalServiceProvider')]() {
            return sp;
        },
        get bsonLibrary() {
            return sp.bsonLibrary;
        },
        aggregate: (...args) => {
            const cursor = sp.aggregate(...args);
            return (0, cursor_wrapper_1.deepInspectCursorWrapper)(cursor);
        },
        aggregateDb: (...args) => {
            const cursor = sp.aggregateDb(...args);
            return (0, cursor_wrapper_1.deepInspectCursorWrapper)(cursor);
        },
        count: forwardedMethod('count', sp),
        estimatedDocumentCount: forwardedMethod('estimatedDocumentCount', sp),
        countDocuments: forwardedMethod('countDocuments', sp),
        distinct: bsonMethod('distinct', sp),
        find: (...args) => {
            const cursor = sp.find(...args);
            return (0, cursor_wrapper_1.deepInspectCursorWrapper)(cursor);
        },
        findOneAndDelete: bsonMethod('findOneAndDelete', sp),
        findOneAndReplace: bsonMethod('findOneAndReplace', sp),
        findOneAndUpdate: bsonMethod('findOneAndUpdate', sp),
        getTopologyDescription: forwardedMethod('getTopologyDescription', sp),
        getIndexes: bsonMethod('getIndexes', sp),
        listCollections: bsonMethod('listCollections', sp),
        readPreferenceFromOptions: forwardedMethod('readPreferenceFromOptions', sp),
        watch: (...args) => {
            const cursor = sp.watch(...args);
            return (0, cursor_wrapper_1.deepInspectCursorWrapper)(cursor);
        },
        getSearchIndexes: bsonMethod('getSearchIndexes', sp),
        runCommand: bsonMethod('runCommand', sp),
        runCommandWithCheck: bsonMethod('runCommandWithCheck', sp),
        runCursorCommand: (...args) => {
            const cursor = sp.runCursorCommand(...args);
            return (0, cursor_wrapper_1.deepInspectCursorWrapper)(cursor);
        },
        dropDatabase: bsonMethod('dropDatabase', sp),
        dropCollection: forwardedMethod('dropCollection', sp),
        bulkWrite: bsonMethod('bulkWrite', sp),
        clientBulkWrite: bsonMethod('clientBulkWrite', sp),
        deleteMany: bsonMethod('deleteMany', sp),
        updateMany: bsonMethod('updateMany', sp),
        updateOne: bsonMethod('updateOne', sp),
        deleteOne: bsonMethod('deleteOne', sp),
        createIndexes: bsonMethod('createIndexes', sp),
        insertMany: bsonMethod('insertMany', sp),
        insertOne: bsonMethod('insertOne', sp),
        replaceOne: bsonMethod('replaceOne', sp),
        initializeBulkOp: forwardedMethod('initializeBulkOp', sp),
        createSearchIndexes: forwardedMethod('createSearchIndexes', sp),
        close: forwardedMethod('close', sp),
        suspend: forwardedMethod('suspend', sp),
        renameCollection: forwardedMethod('renameCollection', sp),
        dropSearchIndex: forwardedMethod('dropSearchIndex', sp),
        updateSearchIndex: forwardedMethod('updateSearchIndex', sp),
        listDatabases: bsonMethod('listDatabases', sp),
        authenticate: forwardedMethod('authenticate', sp),
        createCollection: forwardedMethod('createCollection', sp),
        getReadPreference: forwardedMethod('getReadPreference', sp),
        getReadConcern: forwardedMethod('getReadConcern', sp),
        getWriteConcern: forwardedMethod('getWriteConcern', sp),
        get platform() {
            return sp.platform;
        },
        get initialDb() {
            return sp.initialDb;
        },
        getURI: forwardedMethod('getURI', sp),
        getConnectionInfo: forwardedMethod('getConnectionInfo', sp),
        resetConnectionOptions: forwardedMethod('resetConnectionOptions', sp),
        startSession: forwardedMethod('startSession', sp),
        getRawClient: forwardedMethod('getRawClient', sp),
        createClientEncryption: forwardedMethod('createClientEncryption', sp),
        getFleOptions: forwardedMethod('getFleOptions', sp),
        createEncryptedCollection: forwardedMethod('createEncryptedCollection', sp),
        async getNewConnection(...args) {
            const newSp = await sp.getNewConnection(...args);
            return deepInspectServiceProviderWrapper(newSp);
        },
    };
}
function bsonMethod(key, sp) {
    if (!sp[key])
        return undefined;
    return async function (...args) {
        const result = await sp[key](...args);
        (0, custom_inspect_1.addCustomInspect)(result);
        return result;
    };
}
function forwardedMethod(key, sp) {
    if (!sp[key])
        return undefined;
    return function (...args) {
        return sp[key](...args);
    };
}
