// @mongosh/shell-api/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADMIN_DB = exports.namespaceInfo = exports.asPrintable = exports.shellApiType = exports.CURSOR_FLAGS = exports.ALL_API_VERSIONS = exports.ALL_PLATFORMS = exports.ALL_TOPOLOGIES = exports.ALL_SERVER_VERSIONS = exports.ServerVersions = void 0;
exports.ServerVersions = {
    latest: '999.999.999',
    earliest: '0.0.0',
};
exports.ALL_SERVER_VERSIONS = [
    exports.ServerVersions.earliest,
    exports.ServerVersions.latest,
];
exports.ALL_TOPOLOGIES = [
    'ReplSet',
    'Standalone',
    'Sharded',
    'LoadBalanced',
];
exports.ALL_PLATFORMS = ['Compass', 'Browser', 'CLI'];
exports.ALL_API_VERSIONS = [0, Infinity];
exports.CURSOR_FLAGS = {
    2: 'tailable',
    4: 'SlaveOk',
    8: 'oplogReplay',
    16: 'noTimeout',
    32: 'awaitData',
    64: 'exhaust',
    128: 'partial',
};
exports.shellApiType = Symbol.for('@@mongosh.shellApiType');
exports.asPrintable = Symbol.for('@@mongosh.asPrintable');
exports.namespaceInfo = Symbol.for('@@mongosh.namespaceInfo');
exports.ADMIN_DB = 'admin';
