// @mongosh/shell-api/lib/mongo-errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformMongoErrorPlugin = void 0;
exports.rephraseMongoError = rephraseMongoError;
const ERROR_REPHRASES = [
    {
        code: 13435,
        replacement: (message) => message.includes('db.runCommand')
            ? message
            : 'not primary - consider using db.getMongo().setReadPref() or readPreference in the connection string',
    },
];
function rephraseMongoError(error) {
    if (!isMongoError(error)) {
        return error;
    }
    const e = error;
    const message = e.message;
    const rephrase = ERROR_REPHRASES.find((m) => {
        if (m.matchMessage) {
            return typeof m.matchMessage === 'string'
                ? message.includes(m.matchMessage)
                : m.matchMessage.test(message);
        }
        return m.code !== undefined && e.code === m.code;
    });
    if (rephrase) {
        e.message =
            typeof rephrase.replacement === 'function'
                ? rephrase.replacement(message)
                : rephrase.replacement;
    }
    return e;
}
function isMongoError(error) {
    return /^Mongo([A-Z].*)?Error$/.test(Object.getPrototypeOf(error)?.constructor?.name ?? '');
}
class TransformMongoErrorPlugin {
    transformError(err) {
        return rephraseMongoError(err);
    }
}
exports.TransformMongoErrorPlugin = TransformMongoErrorPlugin;
