// @mongosh/service-provider-core/lib/connect-info.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = getConnectExtraInfo;
const getBuildInfo = __importStar(require("mongodb-build-info"));
function getHostInformation(host) {
    if (!host) {
        return {
            is_localhost: false,
            is_do_url: false,
            is_atlas_url: false,
        };
    }
    if (getBuildInfo.isLocalhost(host)) {
        return {
            is_localhost: true,
            is_do_url: false,
            is_atlas_url: false,
        };
    }
    if (getBuildInfo.isDigitalOcean(host)) {
        return {
            is_localhost: false,
            is_do_url: true,
            is_atlas_url: false,
        };
    }
    return {
        is_localhost: false,
        is_do_url: false,
        is_atlas_url: getBuildInfo.isAtlas(host),
    };
}
function getConnectExtraInfo({ connectionString, buildInfo, atlasVersion, resolvedHostname, isLocalAtlas, serverName = 'unknown', }) {
    const auth_type = connectionString?.searchParams.get('authMechanism') ?? undefined;
    const uri = connectionString?.toString() ?? '';
    buildInfo ??= {};
    const { isDataLake: is_data_federation, dlVersion } = getBuildInfo.getDataLake(buildInfo);
    const { serverOs, serverArch } = getBuildInfo.getBuildEnv(buildInfo);
    const isAtlas = !!atlasVersion?.atlasVersion || getBuildInfo.isAtlas(uri);
    return {
        ...getHostInformation(resolvedHostname || uri),
        is_atlas: isAtlas,
        server_version: buildInfo.version,
        node_version: process.version,
        server_os: serverOs || undefined,
        uri,
        server_arch: serverArch || undefined,
        is_enterprise: getBuildInfo.isEnterprise(buildInfo),
        auth_type,
        is_data_federation,
        is_stream: getBuildInfo.isAtlasStream(uri),
        dl_version: dlVersion || undefined,
        atlas_version: atlasVersion?.atlasVersion ?? null,
        is_genuine: serverName === 'mongodb' || serverName === 'unknown',
        non_genuine_server_name: serverName,
        is_local_atlas: isLocalAtlas,
    };
}
