// @mongosh/logging/lib/logging-and-telemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggingAndTelemetry = void 0;
exports.setupLoggingAndTelemetry = setupLoggingAndTelemetry;
const mongodb_redact_1 = require("mongodb-redact");
const util_1 = require("util");
const mongodb_log_writer_1 = require("mongodb-log-writer");
const mongodb_log_writer_2 = require("mongodb-log-writer");
const devtools_connect_1 = require("@mongodb-js/devtools-connect");
const helpers_1 = require("./helpers");
const stream_1 = require("stream");
function setupLoggingAndTelemetry(props) {
    const loggingAndTelemetry = new LoggingAndTelemetry(props);
    loggingAndTelemetry.setup();
    return loggingAndTelemetry;
}
class LoggingAndTelemetry {
    constructor({ bus, analytics, userTraits, mongoshVersion, deviceId, }) {
        this.pendingBusEvents = [];
        this.pendingTelemetryEvents = [];
        this.isSetup = false;
        this.isBufferingBusEvents = false;
        this.isBufferingTelemetryEvents = false;
        this.setupTelemetryPromise = Promise.resolve();
        this.telemetrySetupAbort = new AbortController();
        this.busEventState = {
            hasStartedMongoshRepl: false,
            apiCallTracking: {
                isEnabled: false,
                apiCalls: new helpers_1.MultiSet(),
                deprecatedApiCalls: new helpers_1.MultiSet(),
            },
            usesShellOption: false,
            telemetryAnonymousId: undefined,
            userId: undefined,
        };
        this.bus = bus;
        this.analytics = analytics;
        this.log = LoggingAndTelemetry.dummyLogger;
        this.userTraits = userTraits;
        this.mongoshVersion = mongoshVersion;
        this.deviceId = deviceId;
    }
    setup() {
        if (this.isSetup) {
            throw new Error('Setup can only be called once.');
        }
        this.isBufferingTelemetryEvents = true;
        this.isBufferingBusEvents = true;
        this.setupTelemetryPromise = this.setupTelemetry();
        this.setupBusEventListeners();
        this.isSetup = true;
    }
    async flush() {
        this.runAndClearPendingBusEvents();
        this.telemetrySetupAbort.abort();
        await this.log.flush();
    }
    async setupTelemetry() {
        this.deviceId = await this.deviceId;
        this.runAndClearPendingTelemetryEvents();
    }
    attachLogger(logger) {
        if (!this.isSetup) {
            throw new Error('Run setup() before setting up the log writer.');
        }
        if (this.log !== LoggingAndTelemetry.dummyLogger) {
            throw new Error('Previously set logger has not been detached. Run detachLogger() before setting.');
        }
        this.log = logger;
        this.runAndClearPendingBusEvents();
        this.bus.emit('mongosh:log-initialized');
    }
    async detachLogger() {
        this.log = LoggingAndTelemetry.dummyLogger;
        await this.flush();
    }
    runAndClearPendingBusEvents() {
        let pendingEvent;
        while ((pendingEvent = this.pendingBusEvents.shift())) {
            pendingEvent();
        }
        this.isBufferingBusEvents = false;
    }
    runAndClearPendingTelemetryEvents() {
        let pendingEvent;
        while ((pendingEvent = this.pendingTelemetryEvents.shift())) {
            pendingEvent();
        }
        this.isBufferingTelemetryEvents = false;
    }
    setupBusEventListeners() {
        const onBus = (event, listener) => {
            this.bus.on(event, ((...args) => {
                if (this.isBufferingBusEvents) {
                    this.pendingBusEvents.push(() => listener(...args));
                    return;
                }
                listener(...args);
            }));
            return this.bus;
        };
        const getTrackingProperties = () => ({
            mongosh_version: this.mongoshVersion,
            session_id: this.log.logId,
        });
        const getUserTraits = () => ({
            ...this.userTraits,
            ...getTrackingProperties(),
            device_id: typeof this.deviceId === 'string' ? this.deviceId : 'unknown',
        });
        const getTelemetryUserIdentity = () => {
            return {
                anonymousId: this.busEventState.telemetryAnonymousId ??
                    this.busEventState.userId,
            };
        };
        const track = (event) => {
            const callback = () => this.analytics.track({
                ...getTelemetryUserIdentity(),
                event: event.name,
                properties: {
                    ...getTrackingProperties(),
                    ...event.payload,
                },
            });
            if (this.isBufferingTelemetryEvents) {
                this.pendingTelemetryEvents.push(callback);
            }
            else {
                callback();
            }
        };
        const identify = () => {
            const callback = () => this.analytics.identify({
                ...getTelemetryUserIdentity(),
                traits: getUserTraits(),
            });
            if (this.isBufferingTelemetryEvents) {
                this.pendingTelemetryEvents.push(callback);
            }
            else {
                callback();
            }
        };
        onBus('mongosh:start-mongosh-repl', (ev) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_002), 'repl', 'Started REPL', ev);
            this.busEventState.hasStartedMongoshRepl = true;
        });
        onBus('mongosh:start-loading-cli-scripts', (event) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_003), 'repl', 'Start loading CLI scripts');
            this.busEventState.usesShellOption = event.usesShellOption;
        });
        onBus('mongosh:connect', (args) => {
            const { uri, resolved_hostname, ...argsWithoutUriAndHostname } = args;
            const connectionUri = uri && (0, mongodb_redact_1.redactConnectionString)(uri);
            const atlasHostname = {
                atlas_hostname: args.is_atlas ? resolved_hostname ?? null : null,
            };
            const properties = {
                ...argsWithoutUriAndHostname,
                ...atlasHostname,
            };
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_004), 'connect', 'Connecting to server', {
                userId: this.busEventState.userId,
                telemetryAnonymousId: this.busEventState.telemetryAnonymousId,
                connectionUri,
                ...properties,
            });
            track({
                name: 'New Connection',
                payload: properties,
            });
        });
        onBus('mongosh:start-session', (args) => {
            const normalizedTimingsArray = Object.entries(args.timings).map(([key, duration]) => {
                const snakeCaseKey = (0, helpers_1.toSnakeCase)(key);
                return [snakeCaseKey, duration];
            });
            const normalizedTimings = Object.fromEntries(normalizedTimingsArray);
            track({
                name: 'Startup Time',
                payload: {
                    is_interactive: args.isInteractive,
                    js_context: args.jsContext,
                    ...normalizedTimings,
                },
            });
        });
        onBus('mongosh:new-user', (newTelemetryUserIdentity) => {
            if (!newTelemetryUserIdentity.anonymousId) {
                this.busEventState.userId = newTelemetryUserIdentity.userId;
            }
            this.busEventState.telemetryAnonymousId =
                newTelemetryUserIdentity.anonymousId;
            identify();
        });
        onBus('mongosh:update-user', (updatedTelemetryUserIdentity) => {
            if (updatedTelemetryUserIdentity.anonymousId) {
                this.busEventState.telemetryAnonymousId =
                    updatedTelemetryUserIdentity.anonymousId;
            }
            else {
                this.busEventState.userId = updatedTelemetryUserIdentity.userId;
            }
            identify();
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_005), 'config', 'User updated');
        });
        onBus('mongosh:error', (error, context) => {
            const mongoshError = error;
            this.log[context === 'fatal' ? 'fatal' : 'error']('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_006), context, `${mongoshError.name}: ${mongoshError.message}`, error);
            if (error.name.includes('Mongosh')) {
                track({
                    name: 'Error',
                    payload: {
                        name: mongoshError.name,
                        code: mongoshError.code,
                        scope: mongoshError.scope,
                        metadata: mongoshError.metadata,
                    },
                });
            }
        });
        onBus('mongosh:write-custom-log', (event) => {
            this.log[event.method]('MONGOSH-SCRIPTS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_054), 'custom-log', event.message, event.attr, event.level);
        });
        onBus('mongosh:globalconfig-load', (args) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_048), 'config', 'Loading global configuration file', args);
        });
        onBus('mongosh:evaluate-input', (args) => {
            if ((0, mongodb_redact_1.shouldRedactCommand)(args.input)) {
                this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_007), 'repl', 'Evaluating input', { input: '<sensitive command used>' });
                return;
            }
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_007), 'repl', 'Evaluating input', args);
        });
        onBus('mongosh:use', (args) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_008), 'shell-api', 'Used "use" command', args);
            track({
                name: 'Use',
            });
        });
        onBus('mongosh:show', (args) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_009), 'shell-api', 'Used "show" command', args);
            track({
                name: 'Show',
                payload: {
                    method: args.method,
                },
            });
        });
        onBus('mongosh:setCtx', (args) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_010), 'shell-api', 'Initialized context', args);
        });
        onBus('mongosh:api-call-with-arguments', (args) => {
            let arg;
            try {
                arg = JSON.parse(JSON.stringify(args));
            }
            catch {
                arg = { _inspected: (0, util_1.inspect)(args) };
            }
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_011), 'shell-api', 'Performed API call', (0, mongodb_redact_1.redact)(arg));
        });
        onBus('mongosh:api-load-file', (args) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_012), 'shell-api', 'Loading file via load()', args);
            if (this.busEventState.hasStartedMongoshRepl) {
                track({
                    name: 'Script Loaded',
                    payload: { nested: args.nested },
                });
            }
            else {
                track({
                    name: 'Script Loaded CLI',
                    payload: {
                        nested: args.nested,
                        shell: this.busEventState.usesShellOption,
                    },
                });
            }
        });
        onBus('mongosh:eval-cli-script', () => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_013), 'repl', 'Evaluating script passed on the command line');
            track({
                name: 'Script Evaluated',
                payload: {
                    shell: this.busEventState.usesShellOption,
                },
            });
        });
        onBus('mongosh:mongoshrc-load', () => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_014), 'repl', 'Loading .mongoshrc.js');
            track({
                name: 'Mongoshrc Loaded',
            });
        });
        onBus('mongosh:mongoshrc-mongorc-warn', () => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_015), 'repl', 'Warning about .mongorc.js/.mongoshrc.js mismatch');
            track({
                name: 'Mongorc Warning',
            });
        });
        onBus('mongosh:crypt-library-load-skip', (ev) => {
            this.log.info('AUTO-ENCRYPTION', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_050), 'crypt-library', 'Skipping shared library candidate', ev);
        });
        onBus('mongosh:crypt-library-load-found', (ev) => {
            this.log.warn('AUTO-ENCRYPTION', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_051), 'crypt-library', 'Accepted shared library candidate', {
                cryptSharedLibPath: ev.cryptSharedLibPath,
                expectedVersion: ev.expectedVersion.versionStr,
            });
        });
        onBus('mongosh-snippets:loaded', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_019), 'snippets', 'Loaded snippets', ev);
        });
        onBus('mongosh-snippets:npm-lookup', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_020), 'snippets', 'Performing npm lookup', ev);
        });
        onBus('mongosh-snippets:npm-lookup-stopped', () => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_021), 'snippets', 'npm lookup stopped');
        });
        onBus('mongosh-snippets:npm-download-failed', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_022), 'snippets', 'npm download failed', ev);
        });
        onBus('mongosh-snippets:npm-download-active', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_023), 'snippets', 'npm download active', ev);
        });
        onBus('mongosh-snippets:fetch-index', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_024), 'snippets', 'Fetching snippet index', ev);
        });
        onBus('mongosh-snippets:fetch-cache-invalid', () => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_025), 'snippets', 'Snippet cache invalid');
        });
        onBus('mongosh-snippets:fetch-index-error', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_026), 'snippets', 'Fetching snippet index failed', ev);
        });
        onBus('mongosh-snippets:fetch-index-done', () => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_027), 'snippets', 'Fetching snippet index done');
        });
        onBus('mongosh-snippets:package-json-edit-error', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_028), 'snippets', 'Modifying snippets package.json failed', ev);
        });
        onBus('mongosh-snippets:spawn-child', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_029), 'snippets', 'Spawning helper', ev);
        });
        onBus('mongosh-snippets:load-snippet', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_030), 'snippets', 'Loading snippet', ev);
        });
        onBus('mongosh-snippets:snippet-command', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_031), 'snippets', 'Running snippet command', ev);
            if (ev.args[0] === 'install') {
                track({
                    name: 'Snippet Install',
                });
            }
        });
        onBus('mongosh-snippets:transform-error', (ev) => {
            this.log.info('MONGOSH-SNIPPETS', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_032), 'snippets', 'Rewrote error message', ev);
        });
        onBus('mongosh:api-call', (ev) => {
            if (!this.busEventState.apiCallTracking.isEnabled)
                return;
            const { apiCalls, deprecatedApiCalls } = this.busEventState.apiCallTracking;
            if (ev.deprecated) {
                deprecatedApiCalls.add({ class: ev.class, method: ev.method });
            }
            if (ev.callDepth === 0 && ev.isAsync) {
                apiCalls.add({ class: ev.class, method: ev.method });
            }
        });
        onBus('mongosh:evaluate-started', () => {
            const { apiCalls, deprecatedApiCalls } = this.busEventState.apiCallTracking;
            this.busEventState.apiCallTracking.isEnabled = true;
            deprecatedApiCalls.clear();
            apiCalls.clear();
        });
        onBus('mongosh:evaluate-finished', () => {
            const { apiCalls, deprecatedApiCalls } = this.busEventState.apiCallTracking;
            for (const [entry] of deprecatedApiCalls) {
                this.log.warn('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_033), 'shell-api', 'Deprecated API call', entry);
                track({
                    name: 'Deprecated Method',
                    payload: {
                        ...entry,
                    },
                });
            }
            for (const [entry, count] of apiCalls) {
                track({
                    name: 'API Call',
                    payload: {
                        ...entry,
                        count,
                    },
                });
            }
            deprecatedApiCalls.clear();
            apiCalls.clear();
            this.busEventState.apiCallTracking.isEnabled = false;
        });
        onBus('mongosh-sp:reset-connection-options', () => {
            this.log.info('MONGOSH-SP', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_040), 'connect', 'Reconnect because of changed connection options');
        });
        onBus('mongosh-editor:run-edit-command', (ev) => {
            this.log.error('MONGOSH-EDITOR', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_047), 'editor', 'Open external editor', (0, mongodb_redact_1.redact)(ev));
        });
        onBus('mongosh-editor:read-vscode-extensions-done', (ev) => {
            this.log.error('MONGOSH-EDITOR', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_043), 'editor', 'Reading vscode extensions from file system succeeded', ev);
        });
        onBus('mongosh-editor:read-vscode-extensions-failed', (ev) => {
            this.log.error('MONGOSH-EDITOR', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_044), 'editor', 'Reading vscode extensions from file system failed', {
                ...ev,
                error: ev.error.message,
            });
        });
        onBus('mongosh:fetching-update-metadata', (ev) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_052), 'startup', 'Fetching update metadata', {
                ...ev,
            });
        });
        onBus('mongosh:fetching-update-metadata-complete', (ev) => {
            this.log.info('MONGOSH', (0, mongodb_log_writer_2.mongoLogId)(1_000_000_053), 'startup', 'Fetching update metadata complete', {
                ...ev,
            });
        });
        (0, devtools_connect_1.hookLogger)(this.bus, {
            debug: (...args) => {
                return this.log.debug(...args);
            },
            info: (...args) => {
                return this.log.info(...args);
            },
            warn: (...args) => {
                return this.log.warn(...args);
            },
            error: (...args) => {
                return this.log.error(...args);
            },
            mongoLogId: (...args) => {
                return this.log.mongoLogId(...args);
            },
        }, 'mongosh', (uri) => (0, mongodb_redact_1.redactConnectionString)(uri));
    }
}
exports.LoggingAndTelemetry = LoggingAndTelemetry;
LoggingAndTelemetry.dummyLogger = new mongodb_log_writer_1.MongoLogWriter('', null, new stream_1.Writable({
    write(chunk, encoding, callback) {
        callback();
    },
}));
